fix: Refuse to run out of date semgrep-core-proprietary

nmote opened the pull request:
Currently, if you install semgrep-core-proprietary, then upgrade Semgrep, then try to run the Pro Engine, Semgrep will use the old version of semgrep-core-proprietary. This often results in crashes. Instead, we should ask users to update semgrep-core-proprietary.

Since the semgrep-core-proprietary binary does not know what version of Semgrep it has been built for, the simplest way to accomplish this is to add a file alongside the binary indicating which version of Semgrep installed it. When we execute semgrep-core-proprietary, we can look at that file. If it's different than the current version of Semgrep, we refuse to run the old binary.

Fixes #8873

Test plan:

Rebase onto the v1.44.0 tag to ensure compatibility with the latest released version of `semgrep-core-proprietary` and run the following manual tests:

- `semgrep scan --pro --config rules.yaml`
  - With no semgrep-core-proprietary installed (nice error message)
  - With semgrep-core-proprietary installed (success)
  - With `pro-installed-by.txt` changed to a different version (nice error message)
  - With `pro-installed-by.txt` removed (nice error message)
- `semgrep scan --pro --config rules.yaml --legacy`
  - With no semgrep-core-proprietary installed (bad error message, but same as before)
  - With semgrep-core-proprietary installed (success)
  - With `pro-installed-by.txt` changed to a different version (bad error message, same as when it's not installed)
  - With `pro-installed-by.txt` removed (same bad error message)
- `semgrep scan --config rules.yaml`
  - With no semgrep-core-proprietary installed (works fine)
  - With semgrep-core-proprietary installed (works fine)
  - With `pro-installed-by.txt` changed to a different version (works fine)
  - With `pro-installed-by.txt` removed (works fine)
- `semgrep ci` with Pro toggle on
  - With no semgrep-core-proprietary installed (downloads and runs)
  - With semgrep-core-proprietary installed (runs)
  - With `pro-installed-by.txt` changed to a different version (downloads and runs)
  - With `pro-installed-by.txt` removed (downloads and runs)
- `semgrep ci` with Pro toggle off
  - With no semgrep-core-proprietary installed (runs)
  - With semgrep-core-proprietary installed (runs)
  - With `pro-installed-by.txt` changed to a different version (runs)
  - With `pro-installed-by.txt` removed (runs)

nmote commented on cli/src/semgrep/engine.py:
```diff
@@ -101,7 +102,27 @@ def default_interfile_timeout(self) -> int:
+    # was installed by the current version of Semgrep.
+    #
+    # See also commands/install.py add_semgrep_pro_version_stamp.
+    def is_correct_pro_version(self) -> bool:
```
If only we knew of a language with a proper type checker, where this wasn't a concern to begin with

nmote commented:
Looks like I need to update the job that tests semgrep pro

nmote commented:
Converted to draft while I figure out the test-pro job

aryx commented on cli/bin/semgrep:
```diff
@@ -55,15 +55,36 @@ class CoreNotFound(Exception):
+    # it takes well under a millisecond to import this.
+    from semgrep import __VERSION__
+    # Duplicate of cli/src/semgrep/semgrep_core.py pro_version_stamp_path
+    stamp_path = core_path.parent / "pro-installed-by.txt"
```
hmm you might want to check how it works with our nonroot docker container. See the end of Dockerfile. We actually move semgrep-core out of its usual place in /home/semgrep/bin/ so we might need to do things also for pro-installed-by.txt

nmote replied:
Yeah, since I'm having trouble getting the CI job running where we do just this, I'm now planning to add an option to `semgrep install-semgrep-pro` which will allow the caller to supply a binary instead of downloading it. that will hopefully help with the docker container too, since we can just call that instead of adding the stamp manually.

nmote replied:
@aryx So, I got all the CI jobs passing, and I think this should work fine with the docker container too. It will put the stamp alongside the `semgrep-core-proprietary` binary, so if it has permission to install that binary, it should have permission to write the stamp as well. Is it optimistic to assume that this is tested in CI? All the tests are green. If it's not, and this breaks something, I'll sort it out tomorrow.

aryx reviewed (commented):
can't wait to distribute just one semgrep binary instead of all those shenanigans ...

tophersmith commented:
Hi @nmote! I'm the EM for the [OP](https://github.com/returntocorp/semgrep/issues/8873), just checking in on this. While it's nice that semgrep won't crash, for CI cases or regular runs, this will now look more like a silent failure (i.e. quit, but without scanning). Is there a thought to allowing us to auto-update when there's a version mismatch? i.e. on version mismatch, call `semgrep install-semgrep-pro` either by default (which feels like a logical stance) or if we pass in a `--auto-update-if-needed` flag

nmote commented:
Hey @tophersmith, that is in fact what happens with `semgrep ci` -- it will now automatically download the updated binary if it is out of date. Are you asking about `semgrep scan`? Currently it will never download a binary, and making that change is out of scope for this PR. I'm open to changing that, but I'd have to ask around to see why that decision was made in the first place before doing so.

tophersmith commented:
@nmote Thanks for the reply. You are correct. I didn't catch the `return None` in `get_binary_path` had the side effect to force a new download. This looks great to me then. (and no I don't have an opinion on `semgrep scan`, we won't be using it). Thanks for working on this!

nmote commented:
Sounds good! You can also take a look at my test plan to see the new behavior in a variety of scenarios.

nmote committed:
fix: Refuse to run out of date semgrep-core-proprietary

Currently, if you install semgrep-core-proprietary, then upgrade
Semgrep, then try to run the Pro Engine, Semgrep will use the old
version of semgrep-core-proprietary. This often results in crashes.
Instead, we should ask users to update semgrep-core-proprietary.

Since the semgrep-core-proprietary binary does not know what version of
Semgrep it has been built for, the simplest way to accomplish this is to
add a file alongside the binary indicating which version of Semgrep
installed it. When we execute semgrep-core-proprietary, we can look at
that file. If it's different than the current version of Semgrep, we
refuse to run the old binary.

Fixes #8873

Test plan:

Rebase onto the v1.44.0 tag to ensure compatibility with the latest
released version of `semgrep-core-proprietary` and run the following
manual tests:

- `semgrep scan --pro --config rules.yaml`
  - With no semgrep-core-proprietary installed (nice error message)
  - With semgrep-core-proprietary installed (success)
  - With `pro-installed-by.txt` changed to a different version (nice
    error message)
  - With `pro-installed-by.txt` removed (nice error message)
- `semgrep scan --pro --config rules.yaml --legacy`
  - With no semgrep-core-proprietary installed (bad error message, but
    same as before)
  - With semgrep-core-proprietary installed (success)
  - With `pro-installed-by.txt` changed to a different version (bad
    error message, same as when it's not installed)
  - With `pro-installed-by.txt` removed (same bad error message)
- `semgrep scan --config rules.yaml`
  - With no semgrep-core-proprietary installed (works fine)
  - With semgrep-core-proprietary installed (works fine)
  - With `pro-installed-by.txt` changed to a different version (works
    fine)
  - With `pro-installed-by.txt` removed (works fine)
`semgrep ci` with Pro toggle on
  - With no semgrep-core-proprietary installed (downloads and runs)
  - With semgrep-core-proprietary installed (runs)
  - With `pro-installed-by.txt` changed to a different version
    (downloads and runs)
  - With `pro-installed-by.txt` removed (downloads and runs)
`semgrep ci` with Pro toggle off
  - With no semgrep-core-proprietary installed (runs)
  - With semgrep-core-proprietary installed (runs)
  - With `pro-installed-by.txt` changed to a different version (runs)
  - With `pro-installed-by.txt` removed (runs)

nmote commented:
I sorted out the failing CI job without any hacks using #9051, so this should be good to go.
